### PR TITLE
Improve WAL concurrency performance by @yzqzss and make it optional

### PR DESCRIPTION
**Pull request by equals215**
Merge @yzqzss feature and make it optional, also use non-committed Index in tests.

### CODE REVIEW

**Comment by equals215** [comment]
@yzqzss I do want your review on this PR to ensure that I didn't break **ANY** of what you did. I was very careful and meticulous to separate both logics but mistakes can happen. Thanks!

**Comment by equals215** [comment]
Oh I broke something, on it

**Comment by equals215** [comment]
Never mind I thought I broke something due to the following log message :
```
2024/08/03 15:17:28 WARN AwaitWALCommited called with commit 0
```
but checking out to your last commit it appears that the log message was already appearing, can you confirm it's normal behaviour?

**Comment by yzqzss** [comment]
> Never mind I thought I broke something due to the following log message :
> 
> ```
> 2024/08/03 15:17:28 WARN AwaitWALCommited called with commit 0
> ```
> 
> but checking out to your last commit it appears that the log message was already appearing, can you confirm it's normal behaviour?

https://github.com/internetarchive/Zeno/blob/5eaefe16703bd0ae29164041ac49702cc4e66571/internal/pkg/queue/enqueue.go#L333-L365

Is it possible that none of the writeItemToFile() was called sometime here?

**Comment by equals215** [comment]
Oh yeah if all items are taken out of the handover nothing will have to be written on disk

**Comment by equals215** [comment]
@yzqzss should be good now

**Comment by equals215** [comment]
When you have some time can you describe how your way of writing to the WAL differs from the "original" ?
I want to document it thoroughly

**Comment by yzqzss** [comment]
Oh, thanks for reminding me, I haven't written a description for previous PR, lol.

Yes, We should document the differences. I have to do other things today, will write it tomorrow. :)

**Comment by equals215** [comment]
Can you confirm it's ready to merge? I might work on the queue today and I want to work with your changes integrated

**Comment by yzqzss** [comment]
image
image

These are two changes needed (code reviews above). the rest LGTM.

**Comment by equals215** [comment]
Will implement and merge later on, we'll catch up on Discord for documentation. Have a nice day 🫡

**Comment by yzqzss** [comment]
You too!

File: internal/pkg/queue/index/manager.go [hunk 1]
```diff
@@ -67,12 +90,29 @@ func NewIndexManager(walPath, indexPath, queueDirPath string) (*IndexManager, er
...
+	// Init WAL commit if enabled
+	if useCommit {
+		im.walCommit = new(atomic.Uint64)
+		im.walCommited = new(atomic.Uint64)
+		im.walNotifyListeners = new(atomic.Int64)
+		im.walCommitedNotify = make(chan uint64)
+		im.WalIoPercent = 10
+		im.WalMinInterval = 10 * time.Millisecond
+		im.walStopChan = make(chan struct{})
+		im.stopChan = make(chan struct{})
```
[the remark this answers is not in the document]
Comment (yzqzss) [reply internal/pkg/queue/index/manager.go]: done

File: internal/pkg/queue/index/manager.go [hunk 2]
```diff
@@ -181,12 +420,20 @@ func (im *IndexManager) Pop(host string) (id string, position uint64, size uint6
...
+	return id, position, size, nil
 }
 
+// Close closes the index manager and performs a final dump of the index to disk.
 func (im *IndexManager) Close() error {
+	slog.Info("Closing index manager")
+	defer slog.Info("Index manager closed")
 	im.dumpTicker.Stop()
 	im.stopChan <- struct{}{}
+	im.walStopChan <- struct{}{}
```
[the remark this answers is not in the document]
Comment (yzqzss) [reply internal/pkg/queue/index/manager.go:432]: done

**Comment by yzqzss** [comment]
> When you have some time can you describe how your way of writing to the WAL differs from the "original" ? I want to document it thoroughly

I wrote it in the description of #107.

**Pull request merged**